### DISCUSSION TITLE
How to update OperationalStatus of a device instance

### DISCUSSION THREAD

**Discussion by girigupt** [edgexfoundry/community#126] [Q&A] [answered]
We have this field OperationalStatus as part of device metadata - the values of which are UP, DOWN, UNKNOWN. My question is how do we update it? Should device service/SDK be updating it based on whether it is able to collect data?

**Comment by sivasain** [comment]
In EdgeX, the OperationalStatus field in the device metadata represents the operational state of a device, indicating whether it is UP, DOWN, or UNKNOWN. The responsibility of updating this field typically lies with the device service or SDK that interacts with the physical device.

The device service or SDK should regularly monitor the connectivity and functionality of the device it manages. Based on the device's availability and the ability to collect data, the device service or SDK can update the OperationalStatus field accordingly.

Here's a general approach to updating the OperationalStatus field:

Monitoring Device Connectivity: The device service or SDK should periodically check the connectivity status of the device. This can involve verifying network connectivity, device responsiveness, or any other relevant indicators.

Collecting Data: If the device service or SDK is responsible for collecting data from the device, it should attempt to gather the necessary data during its regular operations.

Updating OperationalStatus: Based on the results of the connectivity checks and data collection attempts, the device service or SDK can update the OperationalStatus field of the device accordingly. For example:

If the device is responsive and data collection is successful, the OperationalStatus can be set to UP.
If the device is not reachable or data collection fails, the OperationalStatus can be set to DOWN.
If the device service or SDK encounters issues in determining the device's operational state, the OperationalStatus can be set to UNKNOWN.
The frequency of updating the OperationalStatus field can vary based on the specific requirements of your system and the devices involved. It's typically done periodically or triggered by specific events, such as connectivity changes or data collection failures.

It's important to note that the exact implementation details may depend on the specific device service or SDK being used, as well as the integration with EdgeX. The device service or SDK should provide the necessary mechanisms to update the OperationalStatus field as part of its functionality.

**Reply by girigupt** [reply]
Wow! Thanks for the detailed explaination. Does the edgex foundry device SDK update the operational status of device based on success/failure of data collection or any other check? I understand that if the device service itself is down, we will not be able to update the operation status of the device

**Reply by sivasain** [reply, marked as answer]
In EdgeX Foundry, the device SDK itself does not directly update the operational status of a device based on the success or failure of data collection or other checks. The responsibility of updating the operational status typically lies with the device service that interacts with the physical device.

The device SDK is a software development kit that provides a set of tools, libraries, and APIs to facilitate the development of device services in EdgeX. It helps in managing the communication and interaction with devices but does not handle the overall operational status of the device service or individual devices.

The device service, which is developed using the device SDK, is responsible for monitoring the connectivity and functionality of the devices it manages. It can perform data collection, check device responsiveness, and handle other device-specific operations. Based on the results of these checks, the device service can update the operational status of the device using the OperationalStatus field in the device metadata.

Therefore, it is up to the implementation of the device service using the device SDK to determine how and when to update the operational status of a device based on the success or failure of data collection or any other relevant checks. The device service has access to the necessary functions and APIs provided by the device SDK to perform these tasks and update the operational status accordingly.

However, if the device service itself is down or not operational, it won't be able to update the operational status of the device. In such cases, the operational status may remain unchanged until the device service is back up and running.

**Reply by girigupt** [reply]
Thanks again
